### ISSUE TITLE
total=False TypedDict Any error

### ISSUE THREAD

**Issue by GBeauregard** [microsoft/pyright#2883]
**Describe the bug**
When `total=false` in a TypedDict, the single-argument form of `get` widens the type to `Any`

**To Reproduce**
```python
from typing import TypedDict

class MyJSON(TypedDict, total=False):
    streams: list[str]

def returns_json() -> MyJSON: ...

reveal_type(returns_json().get("streams"))  # Any | None
reveal_type(returns_json().get("streams", None))  # list[str] | None
```

**Expected behavior**
```python
reveal_type(returns_json().get("streams"))  # list[str] | None
reveal_type(returns_json().get("streams", None))  # list[str] | None
```

**VS Code extension or command-line**
pyright linux commandline 1.1.210

**Additional context**
Found while investigating https://github.com/microsoft/pyright/issues/2875#issuecomment-1014805733

**Comment by erictraut** [comment]
Hmm, I'm not sure what to do about this one. When processing a `TypedDict`, pyright synthesizes methods for the declared fields. The `get` method is synthesized as an overload with one or two overload signatures per field plus a "catch all" overload that accepts an arbitrary `str` as a field name (to allow for dynamic behaviors).

I recently made a change where I don't synthesize the one-argument form of `get` for fields that are not required. This is not type safe, so the intent was to force you to use the two-argument form. The problem is that the "catch all" overload handles this case, so no error is reported.

In the example above, the synthesized method looks like the following:
```python
@overload
def get(self, k: Literal["streams"], default: _T) -> list[str] | _T: ...
@overload
def get(self, k: str, default: _T = None) -> Any | _T: ...
```

I can think of two solutions here:
1. Add back the single-argument form of `get` for non-required fields. This is what mypy appears to do.
2. Eliminate support for dynamic keys by removing the "catch all".

Of the two, I think option 1 is safer.

Thoughts?

**Comment by GBeauregard** [comment]
Ahhh, hmm. I was getting the widening to `Any` in my strictly typed code without error and only found this by chance when investigating the other behavior change, so even if the intent is good the lack of error means it's not very discoverable and I'm probably not the only one who didn't realize `Any` sneaked into their code.

I suspect you're right that removing the support for dynamic keys would cause more problems than anything so I'd prefer 1 (or else some way to make the discouragement of the one-argument form discoverable, but I suspect that's not easy since this doesn't look to be special-cased currently).

Slightly tangentially related: one thing I wish TypedDict et al had was a way to specify the type of "all other keys". I was deserializing a JSON with hundreds of potential keys and only a few had a type that weren't `str` (but were instead a `Mapping` or `Sequence`), and while I technically knew all the possible keys in advance, my life would have been easier with it. There's an issue somewhere for this in `typing` or `mypy` or something.

**Comment by erictraut** [comment]
OK, I'm going to go with option 1 and revert that part of my recent change.

You also make a good point about `Any` creeping in. I'll change that to an `Unknown` so it doesn't go unnoticed.

**Comment by erictraut** [comment]
Oh, I remember now why I used `Any` rather than `Unknown`. If I use the latter, it triggers a `reportUnknownMemberType` diagnostic when referencing the `get` method. I think I'll need to leave that as `Any`. That's consistent with the `get` method on `dict`.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.211, which I just published. It will also be included in the next release of pylance.
